<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title>The interesting analogy between language and biology | Xuelong An Wang</title> <meta name="author" content="Xuelong An Wang"> <meta name="description" content="comments on the interesting parallels that surface in NLP and computational biology, and how each informs about problems and solutions about each other"> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link href="https://cdn.jsdelivr.net/npm/bootstrap@4.6.1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha256-DF7Zhf293AJxJNTmh5zhoYYIMs2oXitRfBjY+9L//AY=" crossorigin="anonymous"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.21.4/dist/bootstrap-table.min.css"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.4.0/css/all.min.css" integrity="sha256-HtsXJanqjKTc8vVQjO4YMhiqFoXkfBsjBWcX91T1jr8=" crossorigin="anonymous"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/academicons@1.9.1/css/academicons.min.css" integrity="sha256-i1+4qU2G2860dGGIOJscdC30s9beBXjFfzjWLjBRsBg=" crossorigin="anonymous"> <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/jwarby/jekyll-pygments-themes@master/github.css" media="" id="highlight_theme_light"> <link href="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.css" rel="stylesheet"> <link rel="shortcut icon" href="/assets/img/brain-fractal-3.jpg"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://awxlong.github.io/blog/2025/lang-biology/"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/jwarby/jekyll-pygments-themes@master/native.css" media="none" id="highlight_theme_dark"> <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script> <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"><span class="font-weight-bold">Xuelong¬†</span>An Wang</a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about</a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">blog<span class="sr-only">(current)</span></a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">research</a> </li> <li class="nav-item "> <a class="nav-link" href="/projects/">projects</a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv</a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="fas fa-moon"></i> <i class="fas fa-sun"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5"> <div class="row"> <div class="col-sm-3"> <nav id="toc-sidebar" class="sticky-top"></nav> </div> <div class="col-sm-9"> <div class="post"> <header class="post-header"> <h1 class="post-title">The interesting analogy between language and biology</h1> <p class="post-meta">January 19, 2025</p> <p class="post-tags"> <a href="/blog/2025"> <i class="fas fa-calendar fa-sm"></i> 2025 </a> ¬† ¬∑ ¬† <a href="/blog/tag/research"> <i class="fas fa-hashtag fa-sm"></i> research</a> ¬† <a href="/blog/tag/food-for-thought"> <i class="fas fa-hashtag fa-sm"></i> food-for-thought</a> ¬† ¬† ¬∑ ¬† <a href="/blog/category/blog-post"> <i class="fas fa-tag fa-sm"></i> blog-post</a> ¬† </p> </header> <article class="post-content"> <div id="markdown-content"> <h1 id="on-the-interesting-parallels-of-language-and-biology-a-taster">On the interesting parallels of language and biology: a taster</h1> <p>There are interesting parallels between language and biology. As a result, problems surfacing in natural language processing (NLP) tasks, may well inform us about possible problems and solutions in computational biology.</p> <p>The simplest common denominator to observe the parallels concerns the analysis of sequences: algorithms and modelling paradigms like recurrent neural networks and self-attention in transformer that help process a sequence of words to solve a task like text classification can surprisingly work with sequences of genes to classify what disease does an RNA sequence correspond to. This is because at a computational level, a string of nucleotides (adenine (A), thymine (T) or uracil (U) in RNA, cytosine (C), and guanine (G)) can be <a href="https://www.sciencedirect.com/science/article/pii/S2001037021000945" rel="external nofollow noopener" target="_blank">represented as a sequence of letters</a>.</p> <p>Another common point concerns the representation of words and biological entities. It would not suffice to represent a word like ‚Äòairplane‚Äô based solely on this string (i.e. this symbol) so it is common to use a fixed-size embedding vector to denote it. This be extracted from open-source libraries like GLoVE and word2Vec. In parallel, molecules also have expert-coded or learnt <a href="https://pubs.acs.org/doi/epdf/10.1021/acs.jcim.9b00237" rel="external nofollow noopener" target="_blank">molecular fingerprints</a> as vector embeddings which are readily available in RDKit. Another case is that of obtaining embeddings for biological entities. A gene can be treated as a vocabulary token, while a cell can be treated <a href="https://www.biorxiv.org/content/10.1101/2023.04.30.538439v1.full.pdf" rel="external nofollow noopener" target="_blank">as a sentence</a>. Another work on a similar vein is <a href="https://www.biorxiv.org/content/10.1101/2023.09.11.557287v2" rel="external nofollow noopener" target="_blank">Cell2Sentence</a>, where gene expression (single-cell transcriptomics) is treated as text.</p> <h1 id="language-agnostic-grammar-structure-for-language-and-biology">Language-agnostic grammar: structure for language and biology</h1> <p>Inspired by Chomsky‚Äôs linguistics, there is work on learning a language-agnostic tree-like structure to learn embeddings of text, called <a href="https://arxiv.org/abs/2305.05588" rel="external nofollow noopener" target="_blank">Self-Structured AutoEncoder</a>. This tree-like structure combines bottom-up composition of local contextual embeddings with top-down decomposition of fully contextual embeddings to learn in an unsupervised routine. Self-StrAE achieves competitive autoencoding performance with respect to a LLM, BERT-Based variant, denoting its low parameter efficiency that reflects the parsimonious nature of grammar.</p> <p>Given the parallels with language, it is interesting to explore how would it work if Self-StrAE is applied to a DNA, or RNA sequence. What will the learnt structure tell us about the DNA sequence. Learnt embeddings would shed light into sequences of DNA which may have fucntional similarities. The compositions found by Self-StrAE would also find sequence motifs, which are short recurring patterns of DNA that map to specific biological functions like protein docking, because of frequent coocurrence leading to similar embeddings.</p> <h1 id="language-respects-constraints-so-does-biology">Language respects constraints, so does biology</h1> <p>Another interesting parallel are the constraints underlying both generation of sentences and biological entities/processes.</p> <p>Imagine that you want to generate sentences. You have a collection of sentences to train a probabilistic model on. Because we can‚Äôt control what the probabilistic model learns from the given data, how do you make sure you don‚Äôt generate an ungrammatical sentence like ‚Äújogging Octopie went‚Äù, where we have a ‚Äúverb-subject-verb‚Äù structure. The simplest way would be to set up a hidden template like ‚Äúsubject-verb-object‚Äù<sup id="fnref:1"><a href="#fn:1" class="footnote" rel="footnote" role="doc-noteref">1</a></sup> in order to offset portion of the probabilistic model‚Äôs support<sup id="fnref:3"><a href="#fn:3" class="footnote" rel="footnote" role="doc-noteref">2</a></sup> to not put any probability mass to structures other than ‚Äúsubject-verb-object‚Äù. Given this, this will first guarantee that sentences follow the correct grammar rules, and second guide <a href="https://awxlong.github.io/blog/2023/sgd/">learning</a> since the model is not exploring over the whole parameter space, but only that which conform to the supplied template.</p> <p>Consider the following analogous task, given some atoms like hydrogen, carbon, oxygen, among others, you want to train a probabilistic model to predict the angle of rotation when such input atoms are assembled together into a compound. This could be relevant for a 3D reconstruction of an assembled molecule. You have a dataset of different atom-atom compounds with their associated rotation angles. If you only rely on learning, then the resulting probabilistic model may put probability mass on biologically implausible rotation angles, even if that implausible probability mass is negligible. In the end, statistics is just about counting and averaging, meaning that even if you only see that some compound like water has a bond angle of around <a href="http://witcombe.sbc.edu/water/chemistrystructure.html" rel="external nofollow noopener" target="_blank">105 degrees</a>, the model‚Äôs support would still put probability mass in-between 0 or 200 degrees <sup id="fnref:2"><a href="#fn:2" class="footnote" rel="footnote" role="doc-noteref">3</a></sup>. Enumerating all possible rules concerning rotation angles would be impractical, so one can resort to biomedical ontologies.</p> <p>Arguably, this is how DeepMind‚Äôs AlphaFold ensures that proteins generated from DNA input sequences are biologically plausible, as can be observed in the model‚Äôs pipeline. In it, the input sequence first goes through a database search that is subsequently converted into embeddings used to generate the protein structure. Without this first phase, this probabilistic model may output proteins that resemble the ones observed during training, at the risk of following impossible configurations.</p> <figure> <img src="/assets/img/alphafold-pipeline.png" alt="Sorry, an unanticipated error occured and the image can't load." width="100%" height="auto"> <figcaption id="afold"> Structural constraints in the first phase concerning genetic database search and structure database search ensures that rotation angles of bonds in the compounds of the protein are biologically plausible, setting structure in learning. Figure extracted from https://www.nature.com/articles/s41586-021-03819-2 </figcaption> </figure> <p>Generating implausible proteins, or invalid angle rotations between bonds is akin to generating sentences that violate grammar rules.</p> <p>It is debatable whether the rules can be learnt from data. Rather, the role of rules and constraints might be to shape learning.</p> <figure> <img src="/assets/img/nesy-cell.png" alt="Sorry. Image couldn't load." width="100%" height="auto"> <figcaption id="cell-embedder">A depiction of a neurosymbolic programming approach to virtual cell. Consider a single nucleotide sequenceüß¨ changing over time. A neural-network driven search of programs takes the sequence as input, and outputs a sequence of program or steps that transforms the input sequence to a final state. The sequence of steps serve as an explanation to the final state, which could correspond to cancer. </figcaption> </figure> <p>For further thought experiments, imagine whether a model can learn what are the rules of programming languages by only observing code, or infer the rules of grammar from just observing written text.</p> <h1 id="nlp-inspiring-future-biomedical-research-directions">NLP inspiring future biomedical research directions</h1> <p>A dictionary with mostly indisputable grammar rules are akin to biological ontologies like the Gene Ontology An interesting thought experiment hence, is can this grammar be expanded given the text available, analogous to asking whether</p> <figure> <img src="/assets/img/expand-ontology.png" alt="Sorry, an unanticipated error occured and the image can't load." width="100%" height="auto"> <figcaption id="ontology"> . </figcaption> </figure> <h1 id="final-thoughts">Final thoughts</h1> <p>I believe that because we can natively read Human language, research on NLP is more commoditized and well-received by the public than computational genomics, despite the parallels outlined above. I think it is worth for me and anyone interested in computational biology to closely follow the research literature on language modelling to draw inspiration ofr DNA/RNA sequence modelling.</p> <p>Finally, I leave you the following thought experiment: we humans can only read human language. While we can not natively understand the language of biology, our AI-based tools can understand them.</p> <p>‚ÄòAI may just turn out to be the language to describe biology‚Äô - Demis Hassabis‚Äôs <a href="https://www.youtube.com/watch?v=Gfr50f6ZBvo" rel="external nofollow noopener" target="_blank">statement</a></p> <p>If you have answers, share thoughts, you can leave a comment or please email me!</p> <div class="footnotes" role="doc-endnotes"> <ol> <li id="fn:1"> <p>Sentences don‚Äôt follow this simple template, but it helps get the idea across that placing structure into the support of a probabilistic model helps guide learning.¬†<a href="#fnref:1" class="reversefootnote" role="doc-backlink">‚Ü©</a></p> </li> <li id="fn:3"> <p>A probabilistic model‚Äôs support refers to the domain of values for which the output of the model is non-zero.¬†<a href="#fnref:3" class="reversefootnote" role="doc-backlink">‚Ü©</a></p> </li> <li id="fn:2"> <p>While it is true that some bonds between atoms like a carbon-carbon bond have no restricted rotation angle, others like a <a href="https://www.sciencedirect.com/science/article/pii/S0022285217302990" rel="external nofollow noopener" target="_blank">hydrogen peroxide</a> is constrained to a setting of rotation degrees (with some uncertainty).¬†<a href="#fnref:2" class="reversefootnote" role="doc-backlink">‚Ü©</a></p> </li> </ol> </div> </div> </article> <br> <hr> <br> <ul class="list-disc pl-8"></ul> <h2 class="text-3xl font-semibold mb-4 mt-12">Enjoy Reading This Article?</h2> <p class="mb-2">Here are some more articles you might like to read next:</p> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2025/tertulia-kennedy/">Conversatoria sobre la inteligencia artificial en la medicina</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2025/inquery/">InQuery: Text2SQL App coded with LLMs</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2024/manifold-vqa/">Ideas on transitioning from monolith to manifold of models</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2024/thoughts-deepmind-cell/">Thoughts on How to Build the Virtual Cell with Artificial Intelligence</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2024/consilience/">phenotype = (genotype + environment) * consciousness. Reflections on The consilience of knowledge by Edward O. Wilson</a> </li> <div id="disqus_thread" style="max-width: 800px; margin: 0 auto;"> <script type="text/javascript">var disqus_shortname="al-folio",disqus_identifier="/blog/2025/lang-biology",disqus_title="The interesting analogy between language and biology";!function(){var e=document.createElement("script");e.type="text/javascript",e.async=!0,e.src="//"+disqus_shortname+".disqus.com/embed.js",(document.getElementsByTagName("head")[0]||document.getElementsByTagName("body")[0]).appendChild(e)}();</script> <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript" rel="external nofollow noopener" target="_blank">comments powered by Disqus.</a> </noscript> </div> </div> </div> </div> </div> <footer class="fixed-bottom"> <div class="container mt-0"> ¬© Copyright 2025 Xuelong An Wang. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. Photos from <a href="https://unsplash.com" target="_blank" rel="external nofollow noopener">Unsplash</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="https://cdn.jsdelivr.net/npm/bootstrap@4.6.1/dist/js/bootstrap.bundle.min.js" integrity="sha256-fgLAgv7fyCGopR/gBNq2iW3ZKIdqIcyshnUULC4vex8=" crossorigin="anonymous"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@4/imagesloaded.pkgd.min.js"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js"></script> <script defer src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script> <script defer src="https://unpkg.com/bootstrap-table@1.21.4/dist/bootstrap-table.min.js"></script> <script src="/assets/js/no_defer.js?07eaa7cf937a07ef341b1faf1776b8cf"></script> <script defer src="/assets/js/common.js?85108440d5ed580d281b2dcc25e2b2d9"></script> <script defer src="/assets/js/copy_code.js?bd78cf329e9ccb6ed226722e0a87ad8e" type="text/javascript"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script> <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> </body> </html>